pkg/kn/commands/plugin/plugin_verifier.go

```diff
@@ -0,0 +1,259 @@
+	if runtime.GOOS == "windows" {
+		fileExt := strings.ToLower(filepath.Ext(fileInfo.Name()))
+
+		switch fileExt {
+		case ".bat", ".cmd", ".com", ".exe", ".ps1":
+			return eaw
+		}
+		return eaw.addWarning("%s is not executable as it does not have the proper extension", path)
+	}
+
+	mode := fileInfo.Mode()
+	if !mode.IsRegular() && !isSymlink(mode) {
+		return eaw.addWarning("%s is not a file", path)
+	}
+	perms := uint32(mode.Perm())
+
+	var sys *syscall.Stat_t
+	var ok bool
+	if sys, ok = fileInfo.Sys().(*syscall.Stat_t); !ok {
+		// We can check the files' owner/group
+		return eaw.addWarning("cannot check owner/group of file %s", path)
+	}
+
+	isOwner := checkIfUserIsFileOwner(sys.Uid)
+	isInGroup, err := checkIfUserInGroup(sys.Gid)
+	if err != nil {
+		return eaw.addError("cannot get group ids for checking executable status of file %s", path)
+	}
+
+	// User is owner and owner can execute
+	if perms&UserExecute != 0 {
+		if os.Getuid() == 0 {
+			return eaw
+		}
+		if isOwner {
+			return eaw
+		}
+	}
+
+	// User is in group which can execute, but user is not file owner
+	if perms&GroupExecute != 0 {
+		if os.Getuid() == 0 {
+			return eaw
+		}
+		if !isOwner && isInGroup {
+			return eaw
+		}
+	}
+
+	// All can execute, and the user is not file owner and not in the file's perm group
+	if perms&OtherExecute != 0 {
+		if os.Getuid() == 0 {
+			return eaw
+		}
+		if !isOwner && !isInGroup {
+			return eaw
+		}
+	}
+
+	return eaw.addWarning("%s is not executable by current user", path)
+}
+
+func isSymlink(mode os.FileMode) bool {
+	return mode&os.ModeSymlink != 0
+}
+
+func checkIfUserInGroup(gid uint32) (bool, error) {
+	groups, err := os.Getgroups()
+	if err != nil {
+		return false, err
+	}
+	for _, g := range groups {
+		if int(gid) == g {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func checkIfUserIsFileOwner(uid uint32) bool {
+	if int(uid) == os.Getuid() {
+		return true
+	}
+	return false
+}
+
+func (eaw *errorsAndWarnings) addError(format string, args ...interface{}) errorsAndWarnings {
+	eaw.errors = append(eaw.errors, fmt.Sprintf(format, args...))
+	return *eaw
+}
+
+func (eaw *errorsAndWarnings) addWarning(format string, args ...interface{}) errorsAndWarnings {
+	eaw.warnings = append(eaw.warnings, fmt.Sprintf(format, args...))
+	return *eaw
+}
+
+func (eaw *errorsAndWarnings) printWarningsAndErrors(out io.Writer) {
+	printSection(out, "ERROR", eaw.errors)
+	printSection(out, "WARNING", eaw.warnings)
+}
+
+func (eaw *errorsAndWarnings) combinedError() error {
+	if len(eaw.errors) == 0 {
+		return nil
+	}
+	return errors.New(strings.Join(eaw.errors, ","))
+}
+
+func printSection(out io.Writer, label string, values []string) {
+	if len(values) > 0 {
+		printLabelWithConditionalPluralS(out, label, len(values))
+		for _, value := range values {
+			fmt.Fprintf(out, "  - %s\n", value)
+		}
+	}
+}
+
+func printLabelWithConditionalPluralS(out io.Writer, label string, nr int) {
```

Comment:
`...PluralS`... why not `...Plurals`?

Reply:
Because I refer to the plural "S" to be added at the end, and I don't refer to the plural of plural ;-) (which would be "plurals", supposed that something like this exists)

Reply:
OK cool... weird, but cool :)